src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+    * first or second read.  Groups reads in a convenient way for duplicate-marking and
+    * UMI assignment, and ensures that both ends of a read come together.
+    *
+    * Private because it has some serious restrictions!  Only allows:
+    *   - Paired reads only
+    *   - Mapped reads with mapped mates only
+    *   - No secondary or supplementary reads
+    *   - Read and mate _must_ be mapped to the same chromosome
+    * If any of these conditions are violated it will go badly!
+    */
+  private[umi] class EarlierReadComparator extends SAMRecordComparator {
+    override def fileOrderCompare(lhs: SAMRecord, rhs: SAMRecord): Int = compare(lhs, rhs)
+
+    /** A set of functions that extract a value from a read, used in order to compare reads. */
+    private val comparisons: Seq[(SAMRecord,SAMRecord) => Int] = Seq(
+      (lhs, rhs) => lhs.getReferenceIndex - rhs.getReferenceIndex,
+      (lhs, rhs) => sortPosition(lhs) - sortPosition(rhs),
+      (lhs, rhs) => library(lhs).compareTo(library(rhs)),
+      (lhs, rhs) => lhs.getReadName.compareTo(rhs.getReadName),
+      (lhs, rhs) => if (lhs.getFirstOfPairFlag) -1 else 1,
+      (lhs, rhs) => lhs.getFlags - rhs.getFlags
+    )
+
+    /** Compares two reads for sort order. */
+    override def compare(lhs: SAMRecord, rhs: SAMRecord): Int = {
+      // Do some asserting!
+      assertValidRead(lhs)
+      assertValidRead(rhs)
+      comparisons.iterator.map(f => f(lhs, rhs)).find(_ != 0).getOrElse(0)
+    }
+
+    /** Asserts that we didn't get reads we are unable to sort. */
+    final def assertValidRead(rec: SAMRecord) = {
+      assert(rec.getReadPairedFlag,    "Unpaired read: " + rec.getReadName)
+      assert(!rec.getReadUnmappedFlag, "Unmapped read: " + rec.getReadName)
+      assert(!rec.getMateUnmappedFlag, "Read w/unmapped mate: " + rec.getReadName)
+      assert(rec.getReferenceIndex == rec.getMateReferenceIndex, "Read w/mate on different chr: " + rec.getReadName)
+      assert(SAMUtils.getMateCigarString(rec) != null, "Read w/o Mate Cigar tag: " + rec.getReadName)
+      assert(rec.getAttribute("MQ") != null, "Read w/o Mate MQ tag: " + rec.getReadName)
+    }
+
+    /** Retrieves the lower of the rec's and mate's 5' unclipped position for sorting. */
+    final def sortPosition(rec: SAMRecord): Int = {
+      val tmp = rec.getTransientAttribute(SortPositionAttribute)
+      if (tmp != null) {
+        tmp.asInstanceOf[Int]
+      }
+      else {
+        val p1 = if (rec.getReadNegativeStrandFlag) rec.getUnclippedEnd else rec.getUnclippedStart
+        val p2 = if (rec.getMateNegativeStrandFlag) SAMUtils.getMateUnclippedEnd(rec) else SAMUtils.getMateUnclippedStart(rec)
+        val pick = Math.min(p1, p2)
+        rec.setTransientAttribute(SortPositionAttribute, pick)
+        pick
+      }
+    }
+  }
+}
+
+@clp(group=ClpGroups.Umi, description =
+  """
+    |Groups reads together that appear to have come from the same original molecule. Reads
```

Comment:
Not sorted by the outer ends, but by the first based sequenced, which could be inner for non-standard pairings.